### ISSUE TITLE
Give a better look at RubyGems handling the gem directories

### ISSUE THREAD

**Issue by voxik** [rubygems/rubygems#7089]
Looking into #5327 and similar, I believe that RubyGems would deserve second look at where the gems are installed.

First and foremost, RubyGems were historically designed to well support multiple install locations. There essentially were `Gem.default_path` and `Gem.default_dir`. The first was list of all paths where gems were stored and the `Gem.default_dir` was the location, where the gems were installed by default. Please note that these were times, where RubyGems were 3rd partly library and there were no gems shipped as part of Ruby.

Later, RubyGems became part of StdLib and first gems started to be part of Ruby. Unfortunately, the `Gem.default_dir` meaning somehow drifted at this time. While it still was the default installation location, it also become the location of gems which are part of StdLib.

And this is very unfortunate situation. In older days, without bundled and default gems, Fedora was using [this](https://src.fedoraproject.org/rpms/ruby/blob/ad05a88747a3894aa01ad46dd6728d91e8b98e30/f/operating_system.rb) operating_system.rb to configure where to install system wide available gems. Unfortunately, changing this configuration also meant that there were suddenly missing the default/bundled gems. So Fedora adapted and [this](https://src.fedoraproject.org/rpms/ruby/blob/rawhide/f/operating_system.rb) is the current version of operating_system.rb. It uses standard command line options, but these unfortunately has different side effects.

Now back to the context of #5327, can we make sure that:
1) All gem directories are treated the same. Currently, there is essentially single location and using any other location has their issues. This resulted in hackish implementation of default gems, where they could have been just different location.
2) Don't treat user home installed gems any special.
3) Let the vendor configure what is their default location easily.

Maybe the first good step could be to free the StdLib gems from the `Gem.default_dir` and give the `Gem.default_dir` the old meaning, which was just the default install location. Nothing more nothing less.

**Comment by eregon** [comment]
Linking to https://bugs.ruby-lang.org/issues/19972 since it seems highly related.

I replied there, agreed it would be best to store default & bundled gems separate from user-installed gems.

> 2. Don't treat user home installed gems any special.

:+1: I think this would remove a bunch of bugs because user-install is treated quite specially right now IIRC.

**Comment by voxik** [comment]
> Linking to https://bugs.ruby-lang.org/issues/19972 since it seems highly related.

I think this ^^ ticket extracts my last thought (_Maybe the first good step could be to free the StdLib gems from the Gem.default_dir and give the Gem.default_dir the old meaning, which was just the default install location. Nothing more nothing less._) into separate ticket, which is more relevant for Ruby itself. Thx for providing the link here.

**Comment by duckinator** [comment]
I agree that separating things, especially stdlib gems, from `Gem.default_dir` makes sense. We should make sure to this information is exposed for debugging purposes, though &mdash; e.g. as other `Gem.???_dir` functions and in the output of `gem env`.

> user-install is treated quite specially right now 

This actually improved quite a bit recently, because of #7100 (by @voxik). `gem install --user-install` is now just sets `--install-dir` to `Gem.user_dir`.

**Comment by hsbt** [comment]
I rejected the related ruby-core ticket https://bugs.ruby-lang.org/issues/19972 for the same reason.

Now that extracting the standard libraries as default and bundled gems is stabilizing, I don't want to add another directory-separation mechanism and carry its ongoing maintenance cost. `Gem.default_dir` will stay as-is.
